qiskit/primitives/gradient/lin_comb_estimator_gradient.py

```diff
@@ -0,0 +1,224 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Gradient of expectation values with linear combination of unitaries (LCU)
+"""
+
+from __future__ import annotations
+
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Sequence, Type
+
+import numpy as np
+
+from qiskit import transpile
+from qiskit.circuit import (
+    Gate,
+    Instruction,
+    Parameter,
+    ParameterExpression,
+    QuantumCircuit,
+    QuantumRegister,
+)
+from qiskit.circuit.library.standard_gates import (
+    CXGate,
+    CYGate,
+    CZGate,
+    RXGate,
+    RXXGate,
+    RYGate,
+    RYYGate,
+    RZGate,
+    RZXGate,
+    RZZGate,
+)
+from qiskit.quantum_info import Pauli, SparsePauliOp
+
+from ..base_estimator import BaseEstimator
+from ..estimator_result import EstimatorResult
+from .base_estimator_gradient import BaseEstimatorGradient
+
+Pauli_Z = Pauli("Z")
+
+
+@dataclass
+class SubEstimator:
+    coeff: float | ParameterExpression
+    circuit: QuantumCircuit
+    index: int
+
+
+class LinCombEstimatorGradient(BaseEstimatorGradient):
+    """LCU estimator gradient"""
+
+    SUPPORTED_GATES = [
+        "rx",
+        "ry",
+        "rz",
+        "rzx",
+        "rzz",
+        "ryy",
+        "rxx",
+        "cx",
+        "cy",
+        "cz",
+        "ccx",
+        "swap",
+        "iswap",
+        "h",
+        "t",
+        "s",
+        "sdg",
+        "x",
+        "y",
+        "z",
+    ]
+
+    def __init__(
+        self, estimator: Type[BaseEstimator], circuit: QuantumCircuit, observable: SparsePauliOp
+    ):
+        self._circuit = circuit
+        self._observable = observable
+        self._grad, observable = self._preprocessing()
+        circuits = [self._circuit]
+        observables = [self._observable, observable]
+        for param, lst in self._grad.items():
+            for arg in lst:
+                circuits.append(arg.circuit)
+        super().__init__(estimator(circuits=circuits, observables=observables))
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *exc_info):
+        pass
+
+    @classmethod
+    def _gradient_circuits(cls, circuit: QuantumCircuit):
+        circuit2 = transpile(circuit, basis_gates=cls.SUPPORTED_GATES, optimization_level=0)
+        qr_superpos = QuantumRegister(1, "superpos")
+        circuit2.add_register(qr_superpos)
+        circuit2.h(qr_superpos)
+        circuit2.data.insert(0, circuit2.data.pop())
+        circuit2.sdg(qr_superpos)
```

Comment:
Doesn't adding the `sgd` gate yield the imaginary part of the gradients? Afaik, putting only the Hadamard will yield `Re(<psi|H|psi>)` and putting the Hadamard and Sdg `Im(<psi|H|psi>)`. We probably have to make a case distinction here on what users want to evaluate (see also the latest changes in the opflow gradients where you can specify this).

Reply:
`sdg` corresponds to [this line of opflow gradient](https://github.com/Qiskit/qiskit-terra/blob/794db4792e97adca4a78eb4f3411c04e50665df7/qiskit/opflow/gradients/circuit_gradients/lin_comb.py#L550). I picked up only gradient gates required `sdg` (`grad_coeff` is all -0.5j in opflow gradient).

In terms of imaginary part of expectation value, opflow gradient uses `-Y` as superpos operator [line](https://github.com/Qiskit/qiskit-terra/blob/794db4792e97adca4a78eb4f3411c04e50665df7/qiskit/opflow/gradients/circuit_gradients/lin_comb.py#L107). I'm fine to extend the constructor so that users can specify `Z`, `-Y`, or `Z-1jY`.

Reply:
You are both right :) 
Adding only the Hadamard gives `Re(d_θ​<psi(θ​)|H|psi(θ​)>)`. As @t-imamichi mentioned correctly, this gives -- for gates such as single Pauli rotations -- `Re(-0.5j​<psi(θ​)|H|psi(θ​)>)` which is equivalent to `Im(0.5​<psi(θ​)|H|psi(θ​)>)`.
Adding the `aux_meas_op=-Y` from `{Z, -Y, Z-1jY}` allows us to also compute `Im(d_θ​<psi(θ​)|H|psi(θ​)>)`.

Reply:
Thank you for your information. I'll try imag part with [a unit test of gradient framework](https://github.com/Qiskit/qiskit-terra/blob/794db4792e97adca4a78eb4f3411c04e50665df7/test/python/opflow/test_gradients.py#L689).
I would like to make this PR as simple as possible. If you have any suggestion, could you let me know?

Reply:
We could add either a flag to indicate which part of the gradient we want (could be str/enum/bool). If you only want to implement one case I think the real part might be more important, since that's gradient of the energy which we could use in VQE: `d/dθ <psi(θ)|H|psi(θ)> = 2Re(d/dθ​<psi(θ​)|H|psi(θ​)>)`

Reply:
Yes, I'm thinking of enum to represent it. Btw, is there any simple example (circuit and observable) to test the imaginary part?